### DISCUSSION TITLE
Ruby Testing Lessons

### DISCUSSION THREAD

**Discussion by KevinMulhern** [TheOdinProject/curriculum#22455] [Ideas]
### Goals
The testing section in the Ruby course was improved greatly last year with the addition of [Rachels Testing Playground](https://github.com/TheOdinProject/ruby_testing). However the testing lesson content is a bit lacking which is causing a disconnect between the lessons and the exercises.

The main goal of this project will be to improve the testing section with more broken down and comprehensive lessons where we will be able to introduce students to all the important concepts and have them do exercises along the way instead of all the exercises at the end. 

A secondary goal will be to introduce students to testing earlier in the course by moving the testing section between the OOP and Files sections. This will allow students to utilize testing in the intermediary sections of the course and properly aligns the most important things we need to teach students in the Ruby course. 
1: Fundamentals > 2: OOP > 3: Testing  

#### [1: Introduction](https://github.com/TheOdinProject/curriculum/pull/22449) - ✅ 
This lesson explains the reasoning behind automated testing.

* What are automated tests?
* What problems do they solve?

#### 2: RSpec basics - 🚧 
This lesson will introduce students to basic RSpec syntax.

  * Installing RSpec.
  * Structuring test files with describes and it blocks.
  * Anatomy of expectations.
  * lets and before blocks.
  * Exercises: Matcher Exercises from the testing playground.

#### 3: Unit Testing Part 1: Basics - ⌛ 
This lesson will introduce a partially completed project that we will use in the examples.

  * What are unit tests?
  * Testing Query methods.
  * Testing Different Paths through methods with context.
  * Don't test private methods.
  * Exercise: Finish unit testing more of the example project methods.
  * Exercise: Unit test a few ready made value object classes.

#### 4: Unit Testing Part 2: Mocks and Stubs - ⌛ 
This lesson will build off the lesson before it by using the same example.
 
  * What are mocks.
  * What are command methods?
  * What are stubs?
  * How to use them.
  * Disadvantages.
  * Exercise: Finish mocking more of the example projects crud actions to its database.
  * Exercises: Mock a weather service response in a test to make it predicable.
  * Exercises: Mock a quote of the day service response in a test to make it predicable.
  * Exercises: Mock a database service response that takes a while in a test to make the test fast.  

#### 5: End-to-end Testing - ⌛ 
This lesson will once again build off the unit testing lessons by getting the student to write a couple of end-to-end tests for it using [aruba](https://github.com/cucumber/aruba).

* What are end-to-end tests?
* What to test - test the Happy Path and don't try to cover all edge cases.
* The testing Pyramid.
* Exercise: write some more end-to-end tests for the example project. 

#### 6: TDD - 🚧 
This final lesson will introduce the TDD workflow.

* What is TDD and its benefits.
* Explain the red green refactor cycle.
* Walk the student through TDD'ing a couple of methods of a small class.
* Exercise: Finishing TDD'ing a few more methods in the example class used in the content.
* Exercise: TDD a value object.
* Exercise: TDD a class that has a collaborator which doesn't exist yet using mocks.

**Comment by KevinMulhern** [comment]
I've made a good start to the second lesson, so I'm on that at the moment. 

The same person should work on the unit testing and end-to-end testing lessons as they all build off each other. I've already put a good amount of thought into those and have them mostly outlined so I don't mind doing those three as well.

However if anyone wants to help me with these, these are up for grabs:
* The Final TDD lesson. 
* We will need to do some work to the testing playground for the matchers exercises used in the second lesson. This is best left until after the content is written for the second lesson. 
* The exercises for all the other lessons after the content is written 

Drop me a line if you are interested in taking any of these on please 😄

**Reply by wise-king-sullyman** [reply]
I'd like to try and take on the TDD lesson, is there any particular timebox you've got in mind for it?

**Reply by KevinMulhern** [reply]
Nothing too concrete, but I'm optimistically hoping it takes 2 - 3 weeks to finish up the lessons I'm on 😆

**Reply by rlmoser99** [reply]
@KevinMulhern I'd love to help write some of the exercises as the lesson content is written. Let me know when the second lesson is complete and we can discuss your vision for the corresponding exercises.

**Reply by KevinMulhern** [reply]
Will do, thanks @rlmoser99 😄

**Reply by EY-Suhaib-Siddiqui** [reply]
Hi

**Reply by wise-king-sullyman** [reply]
@EY-Suhaib-Siddiqui please reach out to us on our Discord, this isn't the right venue for chatting.

**Comment by KevinMulhern** [comment]
Moves this project to the new system: https://github.com/TheOdinProject/top-meta/projects/4?add_cards_query=is%3Aopen
